transport/ice/IceSession.cpp

```diff
@@ -1091,31 +1270,172 @@ void IceSession::stateCheck(const uint64_t now)
             if (candidatePair->localCandidate.transportType == TransportType::TCP)
             {
-                if (_eventSink)
-                {
-                    _eventSink->onIceDiscardCandidate(this,
-                        candidatePair->localEndpoint.endpoint,
-                        candidatePair->remoteCandidate.address);
-                }
-                removeRemoteCandidate(candidatePair->remoteCandidate);
-                removeCandidatePair(candidatePair);
-            }
-            else if (candidatePair->remoteCandidate.type == IceCandidate::Type::PRFLX)
-            {
-                // port may be recycled in firewall anyway
                 candidatePair->state = ProbeState::Failed;
-                removeRemoteCandidate(candidatePair->remoteCandidate);
+                onCandidatePairRemoved(candidatePair);
+                std::iter_swap(it, --itEnd);
+                continue;
             }
         }
+        // If didn't fail, then it can be a backup in case the nominated becomes unreachable
+        else if (candidatePair->remoteCandidate.type == IceCandidate::Type::PRFLX &&
+            candidatePair->state == ProbeState::Failed)
+        {
+            logger::debug("freezePendingProbes - remove failed candidate: %s %s %s - %s",
+                _logId.c_str(),
+                toString(candidatePair->localEndpoint.endpoint->getTransportType()).c_str(),
+                toString(candidatePair->remoteCandidate.type).c_str(),
+                candidatePair->localCandidate.address.toString().c_str(),
+                candidatePair->remoteCandidate.address.toString().c_str());
+
+            onCandidatePairRemoved(candidatePair);
+            std::iter_swap(it, --itEnd);
+            continue;
+        }
+
+        addCandidateToListIfNotPresent(_remoteCandidates, candidatePair->remoteCandidate);
+        ++it;
+    }
+
+    _candidatePairs.erase(itEnd, _candidatePairs.end());
+}
+
+void IceSession::removeNonReadableRemoteCandidates(uint64_t now)
+{
+    // If we are in the limit, we will be more aggressive on removing candidates
+    // by removing HOST and SRFLX candidates that didn't work or are inactive for long period.
+    // Wen the list is not full, we will only remove PRFLX
+    const bool aggressiveRemove = _remoteCandidates.size() >= _config.maxCandidateCount;
+    _remoteCandidates.clear();
+
+    auto itEnd = _candidatePairs.end();
+
+    for (auto it = _candidatePairs.begin(); it != itEnd;)
+    {
+        auto* candidatePair = it->get();
+
+        // Let's keep the nominated regardless the current state
+        if (candidatePair->nominated)
+        {
+            addCandidateToListIfNotPresent(_remoteCandidates, candidatePair->remoteCandidate);
+            ++it;
+            continue;
+        }
+
+        // Keep HOST and SRFLX if we are not in aggressive removing.
+        if (!aggressiveRemove && candidatePair->remoteCandidate.type != IceCandidate::Type::PRFLX)
```

Comment:
If the candidates list is not full `aggressiveRemove = false`, we will keep SRFLX and HOST candidates